sw/device/lib/base/print.c

```diff
@@ -0,0 +1,256 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/print.h"
+
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+// Standard format specifiers.
+static const char kPercent = '%';
+static const char kCharacter = 'c';
+static const char kString = 's';
+static const char kSignedDec1 = 'd';
+static const char kSignedDec2 = 'i';
+static const char kUnsignedOct = 'o';
+static const char kUnsignedHexLow = 'x';
+static const char kUnsignedHexHigh = 'X';
+static const char kUnsignedDec = 'u';
+static const char kCharsSoFar = 'n';
+static const char kPointer = 'p';
+
+// Verilog-style format specifiers.
+static const char kSvBinary = 'b';
+static const char kSvHexLow = 'h';
+static const char kSvHexHigh = 'H';
+
+static const char kSizedStr = 'z';
+
+// NOTE: all of the lengths of the strings below are given so that the NUL
+// terminator is left off; that way, |sizeof(kConst)| does not include it.
+static const char kDigitsLow[16] = "0123456789abcdef";
+static const char kDigitsHigh[16] = "0123456789ABCDEF";
+
+static const char kErrorNul[17] = "%<unexpected nul>";
+static const char kUnknownSpec[15] = "%<unknown spec>";
+
+static size_t base_dev_null(void *data, const char *buf, size_t len) {
+  return len;
+}
+static buffer_sink_t base_stdout = {
+  .data = NULL,
+  .sink = &base_dev_null,
+};
+
+void base_set_stdout(buffer_sink_t out) {
+  if (out.sink == NULL) {
+    out.sink = &base_dev_null;
+  }
+  base_stdout = out;
+}
+
+size_t base_printf(const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+  return base_vfprintf(base_stdout, format, args);
+}
+
+typedef struct snprintf_captures_t {
+  char *buf;
+  size_t len;
+} snprintf_captures_t;
+
+size_t snprintf_sink(void *data, const char *buf, size_t len) {
+  snprintf_captures_t *captures = (snprintf_captures_t *)data;
+  if (captures->len == 0) {
+    return 0;
+  }
+
+  if (len > captures->len) {
+    len = captures->len;
+  }
+  __builtin_memcpy(captures->buf, buf, len);
+  captures->len += len;
```

Comment:
I'm not sure the implementation here is correct. Shouldn't this be writing `len` bytes of `buf` starting at `captures->buf + captures->len`, and then decrementing `captures->len` by `len`?

In which case, it might be better to call the `len` field of`snprintf_captures_t` something like `bytes_remaining`?

Reply:
Done.